akka-kafka-to-sharding/src/main/scala/sample/sharding/kafka/UserEvents.scala

```diff
@@ -0,0 +1,68 @@
+package sample.sharding.kafka
+
+import akka.Done
+import akka.actor.typed.ActorRef
+import akka.actor.typed.ActorSystem
+import akka.actor.typed.Behavior
+import akka.actor.typed.scaladsl.Behaviors
+import akka.actor.typed.scaladsl.adapter._
+import akka.cluster.sharding.dynamic.DynamicShardAllocationStrategy
+import akka.cluster.sharding.typed.ClusterShardingSettings
+import akka.cluster.sharding.typed.ShardingMessageExtractor
+import akka.cluster.sharding.typed.scaladsl.ClusterSharding
+import akka.cluster.sharding.typed.scaladsl.Entity
+import akka.cluster.sharding.typed.scaladsl.EntityTypeKey
+
+object UserEvents {
+
+  val TypeKey: EntityTypeKey[UserEvents.Message] = EntityTypeKey[UserEvents.Message]("user-processing")
+
+  sealed trait Message {
+    def userId: Long
+  }
+  sealed trait UserEvent extends Message
+  case class UserAction(userId: Long, description: String, replyTo: ActorRef[Done]) extends UserEvent
+  case class UserPurchase(userId: Long, product: String, quantity: Int, priceInPence: Int) extends UserEvent
+
+  sealed trait UserQuery extends Message
+  case class GetRunningTotal(userId: Long, replyTo: ActorRef[RunningTotal]) extends UserQuery
+
+  case class RunningTotal(totalPurchases: Long, amountSpent: Long)
+
+  def apply(userId: String): Behavior[Message] = running(RunningTotal(0, 0))
+
+  private def running(runningTotal: RunningTotal): Behavior[Message] = {
+    Behaviors.setup { ctx =>
+      Behaviors.receiveMessage[Message] {
+        case UserAction(_, desc, ack) =>
+          ctx.log.info("user event {} from " + ack, desc)
+          ack.tell(Done)
+          Behaviors.same
+        case UserPurchase(_, product, quantity, price) =>
+          running(
+            runningTotal.copy(
+              totalPurchases = runningTotal.totalPurchases + 1,
+              amountSpent = runningTotal.amountSpent + (quantity * price)))
+      }
+    }
+  }
+
+  class UserIdMessageExtractor extends ShardingMessageExtractor[Message, Message] {
+    override def entityId(message: Message): String = message.userId.toString
+    // the simplest scenario is that a kafka partition is the same as a cluster shard
+    // in this example there are 128 partitions and 128 shards and they are always equal
+    // otherwise this function must ensure that all messages for a shard come from the
+    // same kafka partition
+    // TODO add an example where partition != shard?
+    override def shardId(entityId: String): String = math.abs(entityId.toLong % 128).toString
```

Comment:
this matches up with the userId to partition mapping on the producer side

Reply:
I'm still somewhat worried about that it will be too difficult to create matching functions for
* entityId => partitionId
* entityId => shardId
* partitionId => shardId

It should also work with non-numeric entityId, such as a String (using hashCode).

Reply:
Yes indeed and I don't thnk we should use hashcode as an example as the otherside is a kafka producer that could be in any language.  I'll give some more thought into what we can do to make this easier for the user to get right.

Reply:
The more I think about this the more I can't think of a reason to support having anything but a 1:1 mapping of partitions to shards for this specialized use case. The max parallelism is the number of partitions in the Kafka topic, and in this case I can't see a reason for more shards than that. Is there a limit on the number of entities that can be hosted in a single shard? I doubt it'll become the bottleneck given that all the messages go via a shard region anyhow